### PR TITLE
4.3: multi turn prompt bot

### PR DESCRIPTION
**Pull request by johnataylor**
- uses mvc api controller
- dependency inject all the things
- add ActivityHandler
- add root dialog
- add custom prompt validator

### CODE REVIEW

File: samples/csharp_dotnetcore/05.multi-turn-prompt/AdapterWithLogging.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+// Licensed under the MIT License.
+
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Builder.Integration.AspNet.Core;
+using Microsoft.Bot.Connector.Authentication;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.BotBuilderSamples
+{
+    public class AdapterWithLogging : BotFrameworkHttpAdapter
```
Comment (yochay) [inline samples/csharp_dotnetcore/05.multi-turn-prompt/AdapterWithLogging.cs]: Maybe something like AdapterWithErrorHandling

File: samples/csharp_dotnetcore/05.multi-turn-prompt/Bots/DialogBasedBot.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            _dialog = dialog;
+            _logger = logger;
+        }
+
+        protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
+        {
+            var dialogSet = new DialogSet(_conversationState.CreateProperty<DialogState>("DialogState"));
+            dialogSet.Add(_dialog);
+
+            var dialogContext = await dialogSet.CreateContextAsync(turnContext, cancellationToken);
```
Comment (yochay) [inline samples/csharp_dotnetcore/05.multi-turn-prompt/Bots/DialogBasedBot.cs]: @johnataylor  can you bring the 'Run' method from sample 42 (or 30?)

Comment (johnataylor) [reply samples/csharp_dotnetcore/05.multi-turn-prompt/Bots/DialogBasedBot.cs]: We discussed just this @Kaiqb was in favor of keeping things flat. Perhaps we can iterate a little on this. The big step here is to move all the dialog logic into a ComponentDialog and inject the resulting class. That is what 42 does too. However, 42 also uses a much more straight forward state model. We tried that approach but then found we wanted to use the UserState for parity with the existing sample.

File: samples/csharp_dotnetcore/05.multi-turn-prompt/AdapterWithLogging.cs (hunk 1 above)
Comment (johnataylor) [reply samples/csharp_dotnetcore/05.multi-turn-prompt/AdapterWithLogging.cs]: OK makes sense. We just copied this class from the other project. I'l rename them together.

File: samples/csharp_dotnetcore/05.multi-turn-prompt/DialogExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+using Microsoft.Bot.Schema;
+
+namespace Microsoft.BotBuilderSamples
+{
+    public static class DialogExtensions
+    {
+        public static async Task Run(this Dialog dialog, ITurnContext<IMessageActivity> turnContext, ConversationState conversationState, CancellationToken cancellationToken)
+        {
+            var dialogSet = new DialogSet(conversationState.CreateProperty<DialogState>("DialogState"));
+            dialogSet.Add(dialog);
```
Comment (JonathanFingold) [inline samples/csharp_dotnetcore/05.multi-turn-prompt/DialogExtensions.cs]: This is kind of cool, but requires the dialog set to be constructed the same way every time. If the dev for some mysterious reason decides to add logic to choose between two different dialogs here, things will not go well.

File: samples/csharp_dotnetcore/01.console-echo/Console-EchoBot.csproj [hunk 4]
```diff
@@ -5,7 +5,7 @@
   <PropertyGroup>
     <OutputType>Exe</OutputType>
     <TargetFramework>netcoreapp2.0</TargetFramework>
-    <CodeAnalysisRuleSet>BotBuilder.ruleset</CodeAnalysisRuleSet>
```
Comment (johnataylor) [inline samples/csharp_dotnetcore/01.console-echo/Console-EchoBot.csproj]: this was not intentional????

Comment (johnataylor) [reply samples/csharp_dotnetcore/01.console-echo/Console-EchoBot.csproj]: we're going to clean up the staging when we have more PRs in there

**Pull request merged**